### PR TITLE
Add API for retrieving the seconds-to-expiry for the token, given a TTL.

### PR DESCRIPTION
**Pull request by janwijbrand**
As discussed in issue #4091 perhaps there could be an API to check the number of seconds before the token expires (against a given TTL). This PR is still work in progress. I'd like to know if the implementation is anywhere near acceptable. If it is, I'll add TTL support to MultiFernet and add documentation. Thanks in advance for your feedback.

### CODE REVIEW

File: src/cryptography/fernet.py [hunk 1]
```diff
@@ -74,6 +74,13 @@ def decrypt(self, token, ttl=None):
         timestamp, data = Fernet._get_unverified_token_data(token)
         return self._decrypt_data(data, timestamp, ttl)
 
+    def ttl(self, token, ttl):
+        current_time = int(time.time())
+        timestamp, data = Fernet._get_unverified_token_data(token)
+        # Verify the token was not tampered with, but do not fail on expiry.
+        self._decrypt_data(data, timestamp, None)
```
Comment (reaperhulk) [inline src/cryptography/fernet.py]: This will, in addition to verifying the HMAC, decrypt the data. While you're throwing that away we should probably just refactor `_decrypt_data` so that these lines (https://github.com/pyca/cryptography/blob/master/src/cryptography/fernet.py#L105-L110) are in a separate method that we can call.

File: tests/test_fernet.py [hunk 2]
```diff
@@ -122,6 +123,51 @@ def test_bad_key(self, backend):
...
+        ts = "1985-10-26T01:20:01"
+        current_time = calendar.timegm(iso8601.parse_date(ts).utctimetuple())
+        monkeypatch.setattr(time, "time", lambda: current_time)
+
+        issued_ts = struct.pack('>Q', calendar.timegm(
+            datetime.datetime(1985, 10, 26, 1, 19, 1).utctimetuple()))
+
+        f = Fernet(base64.urlsafe_b64encode(b"\x00" * 32), backend=backend)
+        with pytest.raises(InvalidToken):
+            token = base64.urlsafe_b64encode(b"\x80" + issued_ts)
```
Comment (reaperhulk) [inline tests/test_fernet.py]: Instead of trying to construct the token manually please just call `encrypt` on Fernet to get a real token.

File: tests/test_fernet.py [hunk 3]
```diff
@@ -122,6 +123,51 @@ def test_bad_key(self, backend):
...
+        current_time = calendar.timegm(iso8601.parse_date(ts).utctimetuple())
+        monkeypatch.setattr(time, "time", lambda: current_time)
+        assert f.ttl(token, ttl=60) == 1
+
+        # Check the TTL two minutes later. We're a minute over due in this
+        # case.
+        ts = "1985-10-26T01:21:00"
+        current_time = calendar.timegm(iso8601.parse_date(ts).utctimetuple())
+        monkeypatch.setattr(time, "time", lambda: current_time)
+        assert f.ttl(token, ttl=60) == 1
```
Comment (reaperhulk) [inline tests/test_fernet.py]: This does not appear to test what the comment says (it is a duplicate of the previous).

**Comment by janwijbrand** [comment]
Thanks for the feedback @reaperhulk. Concerning "why pass a TTL?": you are right, the caller can decide if a new token needs to be handed out based on the age of the current token. So'd rename `def ttl()` to `def age()` then.

I'll address the other comments too.

**Comment by janwijbrand** [comment]
By "just" computing the age, I think the "Also, just to complicate things a bit we actually allow a bit of slop in calculating expiry of a token"-comment is no longer relevant.

File: tests/test_fernet.py (hunk 3 above)
Comment (janwijbrand) [reply tests/test_fernet.py]: Ah, that was a mistake. Fixed it.

File: tests/test_fernet.py (hunk 2 above)
Comment (janwijbrand) [reply tests/test_fernet.py]: Fixed.

File: src/cryptography/fernet.py (hunk 1 above)
Comment (janwijbrand) [reply src/cryptography/fernet.py]: I factored it out to `def _verify_signature()`.

**Comment by janwijbrand** [comment]
I think I processed the comment and feedback. Please let me know if there's anything missing.

File: tests/test_fernet.py [hunk 4]
```diff
@@ -122,6 +122,44 @@ def test_bad_key(self, backend):
         with pytest.raises(ValueError):
             Fernet(base64.urlsafe_b64encode(b"abc"), backend=backend)
 
+    def test_timestamp_ttl_overdue(self, monkeypatch, backend):
```
Comment (reaperhulk) [inline tests/test_fernet.py]: This test doesn't appear to be related to this feature. We should already have this test elsewhere in `test_fernet`.

Comment (janwijbrand) [reply tests/test_fernet.py]: removed duplicate test

File: docs/fernet.rst [hunk 5]
```diff
@@ -80,6 +80,20 @@ has support for implementing key rotation via :class:`MultiFernet`.
...
+        Computes the age of the token. The caller can then decide if the token
+        is about to expire and for example reissue a new token.
+
+        :param bytes token: The Fernet token. This is the result of calling
+                            :meth:`encrypt`.
+        :returns int: The token's age in seconds.
+        :raises cryptography.fernet.InvalidToken: If the ``token``'s signature
+                                                  is invalid, this exception
+                                                  is raised.
+        :raises TypeError: This exception is raised if ``data`` is not
```
[the remark this answers is not in the document]
Comment (janwijbrand) [reply docs/fernet.rst]: fixed.

**Comment by janwijbrand** [comment]
Anything else I could do to refine the PR for it to be merged?

**Comment by janwijbrand** [comment]
Please let me know if I need to amend or additional work for the PR to be mergeable.

**Comment by reaperhulk** [comment]
This mostly looks good. After some discussion we have some additional API changes we'd like to see, but I'll go ahead and cherry-pick this and make the changes.
